### PR TITLE
Support exploring non-default KV stores

### PR DESCRIPTION
**Pull request by kate-goldenring**
Supports exploring non-default KV stores by adding a top level field to specify the label of which store to explore
Stores the user/pass credentials in the app config store instead of in the "default" KV store

image

~Draft as there is currently an issue getting the template to work with variables~ << Thanks @itowlson!

Thank you @karthik2804 for help with the UI!!

To test it, follow updated README instructions

Associated template updates: https://github.com/fermyon/spin-kv-explorer/pull/28

### CODE REVIEW

File: templates/kv-explorer/content/spin.toml
Comment (itowlson) [inline templates/kv-explorer/content/spin.toml]:
```suggestion
{% raw %}kv_credentials = "{{ kv_explorer_user }}:{{ kv_explorer_password }}"{% endraw %}
```

File: templates/kv-explorer/metadata/snippets/component.txt
Comment (itowlson) [inline templates/kv-explorer/metadata/snippets/component.txt]:
```suggestion
{% raw %}kv_credentials = "{{ kv_explorer_user }}:{{ kv_explorer_password }}"{% endraw %}
```

File: templates/kv-explorer/metadata/snippets/component.txt
Comment (itowlson) [inline templates/kv-explorer/metadata/snippets/component.txt]: `foo`?

File: templates/kv-explorer/content/spin.toml
Comment (kate-goldenring) [reply templates/kv-explorer/content/spin.toml]: Thank you so much for helping me with this and getting a better grasp on Liquid template syntax.

File: templates/kv-explorer/metadata/snippets/component.txt
Comment (kate-goldenring) [reply templates/kv-explorer/metadata/snippets/component.txt]: Great eye! This was left over from some testing.

File: templates/kv-explorer/metadata/snippets/component.txt
Comment (kate-goldenring) [inline templates/kv-explorer/metadata/snippets/component.txt]: I updated this to the anticipated release number but the hash is therefore inaccurate. Should the template be updated in a subsequent PR or can we quickly follow this up with an update to the hash?

Comment (itowlson) [reply templates/kv-explorer/metadata/snippets/component.txt]: As there will need to be an update anyway for the hash, it would be good to hold off the template until then, but it may not be worth the amount of picking apart of this PR!  You can also tag at the current HEAD so that Spin users (other than canary users) get the current template until you're ready to release, at which point you move the tag forward... but again, a bit of a faff if the release is going to be almost immediate!

Comment (kate-goldenring) [reply templates/kv-explorer/metadata/snippets/component.txt]: For not being tied to immediacy, i moved the template changes into a separate PR: https://github.com/fermyon/spin-kv-explorer/pull/28

**Comment by kate-goldenring** [comment]
@karthik2804 @radu-matei @calebschoepp, I do not have the permissions in this repo to add reviewers, but your reviews would be helpful given your previous work in the repo.

@karthik2804 and i discussed how in the future we may want one "explorer" for all resources (KV and SQL). For now, this gets KV to links parity

File: explorer/main.go
Comment (karthik2804) [inline explorer/main.go]: I believe this line will be printed twice. Once within the `GetCredentials` function and the bubbled-up error handling here.

File: explorer/main.go
Comment (karthik2804) [inline explorer/main.go]: nit: could this potentially be confusing to local spin users?

File: explorer/main.go
Comment (kate-goldenring) [reply explorer/main.go]: I wasn't sure about this one. I see the second line as a `log`, but logging doesn't work right now for Go, so instead i used a print. The first i see as additional context to the error. If we just do one, I'd prefer the second, since sometime the error is captured by the front end and never displayed in cloud logs.

File: explorer/main.go
Comment (kate-goldenring) [reply explorer/main.go]: These are logs. They are displayed in the cloud logs not in the UI. The error is caught in the front end and the associated message is here: https://github.com/kate-goldenring/spin-kv-explorer/blob/d8b666365e72552dc3734f3f5bba48f6461810a4/explorer/index.html#L314-L315:

```
alert(`Access denied to store with label "${storeLabel}." Ensure the app is linked to the store with this label and that the label has been specified in the Spin manifest (spin.toml).`);
```

Comment (itowlson) [reply explorer/main.go]: They'll be printed to the local console under `spin up` though won't they?

Comment (karthik2804) [reply explorer/main.go]: As @itowlson points out, that is part of the local spin I was trying to point at.

File: explorer/main.go
Comment (karthik2804) [reply explorer/main.go]: I was trying to point at line 76 and then line 239. I also just noticed that the error message would be of the form:

```
Error getting KV explorer credentials from config store: cannot get credentials pair from config store: {some_err}
```

**Comment by radu-matei** [comment]
@kate-goldenring the engineering team on GH should now have write access to the repo (that was an oversight when we moved the repo recently), so you should be able to add reviewers.

File: readme.md
Comment (radu-matei) [inline readme.md:59]: Is this actually enforced by `spin cloud deploy`?

Comment (kate-goldenring) [reply readme.md:59]: Yes, `spin deploy` checks if you have unset variables and prompts you to set them. 

```
$ spin deploy
Error: The application requires values for the following variable(s) which have not been set: kv_explorer_user, kv_explorer_password. Use the --variable flag to provide values.

Learn more at https://developer.fermyon.com/cloud/faq
```

File: explorer/main.go
Comment (kate-goldenring) [reply explorer/main.go]: I see -- that's a great point. I was hyper focused on the Cloud scenario. Are you suggesting we remove this line? :
```
fmt.Printf("Set the 'kv_explorer_user' and 'kv_explorer_password' variables using 'spin cloud variables set'.\n")
```

Comment (radu-matei) [reply explorer/main.go]: Suggestion:

"The [correct variables] for the application are not set. For local testing, you can disable authentication using the [skip auth variable], or set them in you application using [link to setting app variables for all scenarios running Spin]"?

**Comment by radu-matei** [comment]
Not sure it's caused by this PR, but I'm noticing it's not impossible to skip checking the auth when running locally — so we either want to remove it (and update the docs), or decide if it's still valuable?

File: explorer/main.go
Comment (kate-goldenring) [reply explorer/main.go]: Thanks! Removed the line to avoid the error duplication

File: explorer/main.go
Comment (kate-goldenring) [reply explorer/main.go]: I updated the log to give directions for both local and deployed apps:
```
The 'kv_explorer_user' and 'kv_explorer_password' variables for the application are not set. For deployed applications, set the variables using 'spin cloud variables set'. For local development, you can disable authentication using '--env SPIN_APP_KV_SKIP_AUTH=1' or set them in the application using runtime configuration (https://developer.fermyon.com/spin/v2/dynamic-configuration#application-variables-runtime-configuration)
```

**Comment by kate-goldenring** [comment]
>Not sure it's caused by this PR, but I'm noticing it's not impossible to skip checking the auth when running locally — so we either want to remove it (and update the docs), or decide if it's still valuable?

@Radu, it was caused by this PR. Thanks for catching that. Just pushed a fix

**Review by radu-matei — APPROVED** [review]

**Pull request merged**
